### PR TITLE
chore(codecs): Implement `length_delimited` framing

### PR DESCRIPTION
**Pull request by pablosichert**
Part of #7463, has been split off from #8236.

### CODE REVIEW

**Review by bruceg — COMMENTED** [review]
The code looks good enough (noting @lukesteensen's suggestion), but I wonder about the naming. This is just one form of length-delimited framing, specifically 32-bit binary MSB. There are other ways of doing length delimiting, so a more specific name may be be appropriate.

**Comment by pablosichert** [comment]
> [...] I wonder about the naming. This is just one form of length-delimited framing, specifically 32-bit binary MSB. There are other ways of doing length delimiting, so a more specific name may be be appropriate.

@bruceg good point. Alternatively, we could leave the generic naming and (later) add options to the codec to control the length header. What do you think?

File: src/codec/framers/length_delimited.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+impl Decoder for LengthDelimitedCodec {
+    type Item = Bytes;
+    type Error = BoxedFramingError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.0
+            .decode(src)
+            .map(|bytes| bytes.map(Into::into))
```
[the remark this answers is not in the document]
Comment (pablosichert) [reply src/codec/framers/length_delimited.rs]: Changed in 2a8eafddcbc39806a1617125394b05282f52b432

File: src/codec/framers/length_delimited.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            .map_err(Into::into)
+    }
+
+    fn decode_eof(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.0
+            .decode(src)
+            .map(|bytes| bytes.map(Into::into))
```
[the remark this answers is not in the document]
Comment (pablosichert) [reply src/codec/framers/length_delimited.rs]: Changed in 2a8eafddcbc39806a1617125394b05282f52b432

File: src/codecs/framers/length_delimited.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    fn clone(&self) -> Self {
+        // This is an awful implementation for `Clone` since it resets the
+        // internal state. However, it works for our use case because we
+        // generally only clone a codec that has not been mutated yet.
+        //
+        // Ideally, `tokio_util::codec::LengthDelimitedCodec` should implement
+        // `Clone` and it doesn't look like it was a deliberate decision to
+        // leave out the implementation. All of its internal fields implement
+        // `Clone`, so adding an implementation for `Clone` could be contributed
+        // to the upstream repo easily by adding it to the `derive` macro.
```
Comment (spencergilbert) [inline src/codecs/framers/length_delimited.rs:47]: Did we open an issue around this already?

File: src/codecs/framers/length_delimited.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            .map(|bytes| bytes.map(BytesMut::freeze))
+            .map_err(Into::into)
+    }
+
+    fn decode_eof(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        self.0
+            .decode(src)
+            .map(|bytes| bytes.map(BytesMut::freeze))
+            .map_err(Into::into)
+    }
```
Comment (spencergilbert) [inline src/codecs/framers/length_delimited.rs:68]: I might be reading poorly, but what's the difference between these two? They look identical aside from naming?

**Review by spencergilbert — APPROVED** [review]

File: src/codecs/framers/length_delimited.rs (hunk 4 above)
Comment (pablosichert) [reply src/codecs/framers/length_delimited.rs:68]: Whew, my copy pasta failed me there! Should have been `decode_eof` in the second one...

**Review by spencergilbert — APPROVED** [review]

File: src/codecs/framers/length_delimited.rs (hunk 3 above)
Comment (pablosichert) [reply src/codecs/framers/length_delimited.rs:47]: https://github.com/tokio-rs/tokio/pull/4089

**Review by bruceg — APPROVED** [review]

**Pull request merged**